server/methods/core/groups.js

```diff
@@ -0,0 +1,207 @@
+
+    const matchQuery = { $elemMatch: { shopId: shopId, names: group.name } };
+    const affectedUsers = Accounts.find({ groups: matchQuery }).fetch();
+    let error;
+
+    if (permissionsChanged) {
+      error = updateUsersPermissions(affectedUsers, group, newGroupData, shopId);
+    }
+
+    // update group names on user account
+    error = updateUsersGroupName(affectedUsers, shopId);
+
+    if (!error) {
+      return { status: 200 };
+    }
+    Logger.error(error);
+    throw new Meteor.Error(500, "Update not successful");
+  },
+  /**
+   * group/addUser
+   * @summary adds a user to a permission group for a shop
+   * It updates the user's list of permissions/roles with the defined the list defined for the group
+   * It uses the local *updateUsers-* by passing one user as an array
+   * @param {String} userId - current data of the group to be updated
+   * @param {Object} groupData - info about the group
+   * @param {String} shopId - permissions of the group
+   * @return {Object} - object.status of 200 on success or Error object on failure
+   */
+  "group/addUser": function (userId, groupData, shopId) {
+    check(userId, String);
+    check(groupData, Object);
+    check(groupData.name, String);
+    check(groupData.permissions, [String]);
+    check(shopId, String);
+
+    const user = Accounts.findOne({ _id: userId });
+
+    try {
+      Meteor.call("accounts/addUserPermissions", user.userId, groupData.permissions, shopId);
+      // call this in the new fashion
+      updateUsersGroupName([user], shopId);
+      return { status: 200 };
+    } catch (error) {
+      Logger.error(error);
+      throw new Meteor.Error(500, "Could not add user");
+    }
+  },
+
+  /**
+   * group/removeUser
+   * @summary removes a user from a group for a shop, and updates the user's permission list
+   * @param {String} userId - current data of the group to be updated
+   * @param {String} groupName - name of the group
+   * @param {String} shopId - permissions of the group
+   * @return {Object} - object.status of 200 on success or Error object on failure
+   */
+  "group/removeUser": function (userId, groupName, shopId) {
+    check(userId, String);
+    check(groupName, String);
+    check(shopId, String);
+
+    const user = Accounts.findOne({ _id: userId });
+
+    if (!user) {
+      throw new Meteor.Error(404, "Could not find user");
+    }
+
+    const permissions = getGroupPermissions(Shops.findOne({ _id: shopId }).groups, groupName);
+    try {
+      // delete all permissions belonging to that group
+      updateUsersPermissions([user], { permissions }, {}, shopId);
+      // then re-populate for remaining groups
+      updateUsersGroupName([user], shopId);
+      return { status: 200 };
+    } catch (error) {
+      Logger.error(error);
+      throw new Meteor.Error(500, "Could not add user");
+    }
+  }
+});
+
+function updateUsersGroupName(affectedUsers, shopId) {
```

Comment:
This function seems really expensive. Maybe premature to optimize, but might be worth a conversation about if there's a better way to assign / check group membership